### PR TITLE
google groups module with testing

### PR DESCRIPTION
**Pull request by murphgrainger**

### CODE REVIEW

**Comment by murphgrainger** [comment]
let's go over this PR on a quick call - want to confirm a couple things about each function like:

**Comment by murphgrainger** [comment]
let's go over this PR on a call. i want to comment on a few parts of the code: example - do we need to get all group members ever or are we just doing that to see if someone is part of the group already? if so - there's a faster way to do it.  also want to confirm the remove all method is what you're picturing. also want to quickly chat through the tests and why there are so many diff error scenarios. 👍

File: googleGroups/index.js [hunk 1]
```diff
@@ -1,24 +1,78 @@
...
-// Write a student to a group
-exports.addGroupMember = async () => {
+const scopes = ['https://www.googleapis.com/auth/admin.directory.group'];
+const jwt = new google.auth.JWT(process.env.GOOGLE_ADMIN_CLIENT_EMAIL, null, process.env.GOOGLE_ADMIN_CLIENT_KEY, scopes, 'paola@galvanize.com');
 
-};
-
-// Delete a student from a group
-exports.removeGroupMember = async () => {
+const auth = async () => {
```
Comment (anthonypecchillo) [inline googleGroups/index.js:9]: ### Function Name

Let's use `authenticate` here.

- "To authenticate" - The infinitive form of a verb. 👍 
- "auth" - Is a thing. A process. A noun. 🙅‍♂️

File: googleGroups/index.js [hunk 2]
```diff
@@ -1,24 +1,78 @@
...
-exports.removeAllGroupMembers = async () => {
+// Read all members in a group
+exports.getAllGroupMembers = async (groupId) => {
+  try {
+    const service = await auth();
+    const res = await service.members.list({
+      groupKey: groupId,
+    });
+    if (!res.data.members) return [];
+    return res.data.members;
```
Comment (anthonypecchillo) [inline googleGroups/index.js:28]: This looks like a great place to use a ternary operator.

```javascript
return res.data.members ? res.data.members : [];
```

Or, even cleaner, we could use the "poor-man's (poor-person's?) if":

```javascript
return res.data.members || [];
```

File: googleGroups/index.js [hunk 3]
```diff
@@ -1,24 +1,78 @@
...
+  } catch (error) {
+    return error.message;
+  }
+};
 
+// Add a student to a group
+exports.addGroupMember = async (groupId, userEmail) => {
```
Comment (anthonypecchillo) [inline googleGroups/index.js:35]: ### Function Name

It's impossible for you to have had insight into my intentions here, but can we rename this to `addStudentToGroup`?

Eventually, we may want to implement an `addStaffMemberToGroup` function where we give those group members enhanced permissions.

File: googleGroups/index.js [hunk 4]
```diff
@@ -1,24 +1,78 @@
...
+  try {
+    const service = await auth();
+    const res = await service.members.insert({
+      groupKey: groupId,
+      resource: { email: userEmail },
```
Comment (anthonypecchillo) [inline googleGroups/index.js:40]: Can we also change `userEmail` to `email`?

File: googleGroups/index.js [hunk 5]
```diff
@@ -1,24 +1,78 @@
...
+    return res.data && res.data.status === 'ACTIVE';
+  } catch (error) {
+    return error.message;
+  }
 };
 
-// Read all members in a group
-exports.getAllGroupMembers = async () => {
+// Delete a student from a group
+exports.removeGroupMember = async (groupId, userEmail) => {
```
Comment (anthonypecchillo) [inline googleGroups/index.js:49]: This function name can stay the same, because we can use it for both students and staff members.

File: googleGroups/index.js [hunk 6]
```diff
@@ -1,24 +1,78 @@
...
+    const service = await auth();
+    const members = await exports.getAllGroupMembers(groupId);
+    if (typeof members === 'string') throw new Error(members);
+    await members.forEach((user) => {
+      service.members.delete({
+        groupKey: groupId,
+        memberKey: user.email,
+      });
+    });
+    return true;
```
Comment (anthonypecchillo) [inline googleGroups/index.js:74]: ### Some thoughts

- `members.forEach` is not asynchronous, so I don't think we can `await` it.
- I'm afraid this may `return true` before every async call to `service.members.delete` resolves.

💡Instead, let's use `map`, collect all of the promises in the output array, and then `Promise.all` them, then make sure each deletion was successful before returning `true`, like you do so elegantly in the Salesforce PR.

File: googleGroups/index.js [hunk 7]
```diff
@@ -1,24 +1,78 @@
...
+  } catch (error) {
+    return error.message;
+  }
+};
 
+// Delete all students from a group
+exports.removeAllGroupMembers = async (groupId) => {
+  try {
+    const service = await auth();
+    const members = await exports.getAllGroupMembers(groupId);
```
Comment (anthonypecchillo) [inline googleGroups/index.js:66]: ### Woops!

This is my fault for making every function shell immediately attached to `exports`. Generally we shouldn't call `exports.methodName` in our functions, its just a bit of a code smell I think.

But, I forced you to, and I'm glad you re-used this function here. We can change it in a future clean up PR.

### Question

You mentioned there might be a faster way to accomplish this than by re-using `getAllGroupMembers` first. Is there a way to send a request to the API with the `groupKey` that deletes every member inside the group, without having to supply the group members to delete?

Even better, can we also provide a permissions level and only delete the members who don't have admin privileges?

Am I getting greedy? lol

File: salesforce/index.js [hunk 8]
```diff
@@ -1,15 +1,62 @@
...
+const { SFDC_OPPTY_RECORD_ID, SFDC_SELECT_QUERY } = require('../constants');
+
+const conn = new jsforce.Connection({ loginUrl: process.env.SFDC_LOGIN_URL });
 // Salesforce API Integrations
 // ------------------------------
 
-// Read a sales force report
-exports.getReport = async () => {
-
+const login = async () => {
```
Comment (anthonypecchillo) [inline salesforce/index.js:9]: We'll have to run `git pull --rebase upstream master`:

- In this branch
- After merging the Salesforce PR first
- Before merging this PR

Since this code is actually from the Salesforce PR.

File: googleGroups/googleGroups.test.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  });
+};
+
+const removeUser = async () => {
+  const service = await auth();
+  await service.members.delete({
+    groupKey: GOOGLE_GROUP_ID,
+    memberKey: GOOGLE_GROUP_USER,
+  });
+};
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:56]: We already have functions for this from the module that we are testing! 🙂 

Let's simply:

- Test the methods we intend to re-use _first_
- Test the methods that depend on them _afterward_ 

Then we can keep our code nice and DRY.

File: googleGroups/googleGroups.test.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+const auth = async () => {
+  await jwt.authorize((err, token) => {
+    if (err) return err;
+    return token;
+  });
+
+  return google.admin(
+    { version: 'directory_v1', auth: jwt },
+  );
+};
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:25]: Same as comment on line 42!

File: googleGroups/googleGroups.test.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  const service = await auth();
+  const members = await service.members.list({
+    groupKey: GOOGLE_GROUP_ID,
+  });
+  if (members.length) {
+    await members.forEach((user) => {
+      service.members.delete({
+        groupKey: GOOGLE_GROUP_ID,
+        memberKey: user.email,
+      });
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:37]: I'm afraid we cannot `await` a `forEach` loop. The actual async thing here is `service.members.delete`.

Take a look at my similar comment in the `googleGroups/index.js` file to resolve this one.

File: googleGroups/googleGroups.test.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    const members = await getAllGroupMembers('*****');
+    expect(members).toBe('Resource Not Found: groupKey');
+  });
+  test('Should return an error if no access to group id', async () => {
+    const members = await getAllGroupMembers('googlegroups@gmail.com');
+    expect(members).toBe('Not Authorized to access this resource/api');
+  });
+});
+
+describe('addGroupMember', () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:78]: Match string to new function name, as requested in `googleGroups/index.js`

File: googleGroups/googleGroups.test.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+  test('Should return true if user successfully is added', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:79]: "Should return true if user is successfully added to Google Group"

File: googleGroups/googleGroups.test.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+const removeUser = async () => {
+  const service = await auth();
+  await service.members.delete({
+    groupKey: GOOGLE_GROUP_ID,
+    memberKey: GOOGLE_GROUP_USER,
+  });
+};
+
+describe('getAllGroupMembers', () => {
+  test('Should return an empty array if no members', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:59]: "Should return an empty array if there are no members in the Google Group"

Be very explicit. Remember, one day someone is going to read this test after we leave the company, and we don't want to leave them any room to make any false assumptions that might end up slowing them down.

File: googleGroups/googleGroups.test.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    const members = await getAllGroupMembers(GOOGLE_GROUP_ID);
+    expect(members).toHaveLength(0);
+  });
+  test('Should return an array of group members', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:63]: Can we test more deeply here? Maybe the test should state:

"Should return an array of group member objects if there is at least one member in the Google Group"

Then, we can test:

1. Is `members` an array?
2. Does `members` have a length of 1?
3. Is the first item in `members` an object?
4. Does the first item in `members` have the properties we expect?

You can stack several `expect` assertions in the same `test` callback.

File: googleGroups/googleGroups.test.js [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    await addUser(GOOGLE_GROUP_ID, GOOGLE_GROUP_USER);
+    const members = await getAllGroupMembers(GOOGLE_GROUP_ID);
+    expect(members).toHaveLength(1);
+  });
+  test('Should return an error if group id does not exist', async () => {
+    const members = await getAllGroupMembers('*****');
+    expect(members).toBe('Resource Not Found: groupKey');
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:70]: 👍

File: googleGroups/googleGroups.test.js [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  });
+  test('Should return an error if no access to group id', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:72]: 👍 👍

File: googleGroups/googleGroups.test.js [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+  });
+  test('Should return an error if user is already in group', async () => {
+    const response = await addGroupMember(GOOGLE_GROUP_ID, GOOGLE_GROUP_USER);
+    expect(response).toBe('Member already exists.');
+  });
+  test('Should return an error if user does not exist', async () => {
+    const response = await addGroupMember(GOOGLE_GROUP_ID, '****@gmail.com');
+    expect(response).toContain('Resource Not Found');
+  });
+  test('Should return an error if invalid key provided', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:92]: As a reader who knows nothing about the Google Admin SDK, I don't know what you mean by "key" here.

Because "key" is such a general term, maybe we can be more specific. I see a thing called `memberKey` in the next test. Is this what you're referring to? Can we use that name instead?

File: googleGroups/googleGroups.test.js [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+  });
+  test('Should return an error if user was already not in group', async () => {
+    const response = await removeGroupMember(GOOGLE_GROUP_ID, GOOGLE_GROUP_USER);
+    expect(response).toBe('Resource Not Found: memberKey');
+  });
+  test('Should return an error if user does not exist', async () => {
+    const response = await removeGroupMember(GOOGLE_GROUP_ID, '****@gmail.com');
+    expect(response).toContain('Resource Not Found');
+  });
+  test('Should return an error if invalid key provided', async () => {
```
Comment (anthonypecchillo) [inline googleGroups/googleGroups.test.js:111]: "key" => "memberKey"

File: googleGroups/index.js (hunk 4 above)
Comment (murphgrainger) [reply googleGroups/index.js:40]: will change! for context since the function takes a groupId (which is also the group email) and a user's email i didn't want it to be confusing

File: googleGroups/googleGroups.test.js [hunk 20]
```diff
@@ -61,9 +62,11 @@ describe('getAllGroupMembers', () => {
     expect(members).toHaveLength(0);
   });
   test('Should return an array of group members', async () => {
-    await addUser(GOOGLE_GROUP_ID, GOOGLE_GROUP_USER);
+    await addUser();
```
Comment (murphgrainger) [inline googleGroups/googleGroups.test.js:65]: @anthonypecchillo this is the test that is failing. if i log the responses from each function - it is adding the user successfully before the test, then looking for all the users, but still getting an empty array. my last thought is to set a timeout to wait for the fetch all users so there is time for the add to "settle" but i thought doing it async/await should achieve that already?

File: googleGroups/googleGroups.test.js (hunk 15 above)
Comment (murphgrainger) [reply googleGroups/googleGroups.test.js:64]: added some of this but think you might want it more verbose. to me if i check toHaveLength that is also checking if it's an array. could be a string and pass but if i also have expect(members[0]) it would have to be an array or it would fail?
